### ISSUE TITLE
Final not honored when applied post-hoc

### ISSUE THREAD

**Issue by avylove** [python/mypy#20144]
**Bug Report**

The behavior of ``Final`` is a bit inconsistent. It can be applied post-hoc, but then it won't be honored.

**To Reproduce**

```python
import platform
import sys

from typing import Final

SERIES_STD = ' ▌█' if platform.system() == 'Windows' else ' ▏▎▍▌▋▊▉█'
if sys.__stdout__ and sys.__stdout__.encoding:
    try:
        SERIES_STD.encode(sys.__stdout__.encoding)
    except UnicodeEncodeError:
        SERIES_STD = ' |'

SERIES_STD: Final = SERIES_STD
SERIES_STD = '| '
```

**Expected Behavior**

I expect an assignment error for the final line to raise an error.

```console
error: Cannot assign to final name "SERIES_STD"  [misc]
```

Some people might argue applying Final post-hoc itself should raise an error, but I'm not one of them.

**Actual Behavior**
```console
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used:  1.18.2
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): No Config
- Python version used: 3.13.9

**Comment by sterliakov** [comment]
I think `SERIES_STD: Final = SERIES_STD` should trigger a diagnostic `[no-redef]`, this is no better than

```python
x = 0
x: int = 1
```

This must have something to do with us stripping `Final` early after setting the flag.

**Comment by avylove** [comment]
> I think `SERIES_STD: Final = SERIES_STD` should trigger a diagnostic `[no-redef]`

This is what I initially expected to happen. I don't love it, but it is what is specified in  [PEP 591](https://peps.python.org/pep-0591/)

> The two main rules for defining a final name are:
> 
> - There can be at most one final declaration per module or class for a given attribute. There can’t be separate class-level and instance-level constants with the same name.
> 
> - There must be exactly one assignment to a final name
